### DISCUSSION TITLE
mini.statusline: How to Enable icons for Warning, Error, Hint, etc. ?

### DISCUSSION THREAD

**Discussion by mezeirobert0** [nvim-mini/mini.nvim#1177] [Q&A] [answered]
To avoid repeating myself, I will link my Reddit post on Neovim sub:
https://www.reddit.com/r/neovim/comments/1f1lafg/need_to_help_with_customizing_ministatusline/

**Comment by echasnovski** [comment, marked as answer]
Customizing statusline content is expected by [creating a function](https://github.com/echasnovski/mini.nvim/blob/2c55015a518e11ea9d49bb96078f5ede51f5d2af/doc/mini-statusline.txt#L9-L13) and setting it to `config.content.active`. Like this:

```lua
local my_active_content = function()
  -- Your code goes here
end
require('mini.statusline').setup({ content = { active = my_active_content } })
```

Here is an [example function used for default content](https://github.com/echasnovski/mini.nvim/blob/2c55015a518e11ea9d49bb96078f5ede51f5d2af/doc/mini-statusline.txt#L87-L105). Copy paste it as `my_active_content` body as a start and tweak it until you have your final result.

------

As you pointed in Reddit post, indeed using `signs` argument in `MiniStatusline.section_diagnostics()` is the suggested way to customize its signs. So it will be something like this:

```lua
local diag_signs = { ERROR = '!', WARN = '?', INFO = '@', HINT = '*' }
local my_active_content = function()
  -- ...
  local diagnostics = MiniStatusline.section_diagnostics({ trunc_width = 75, signs = diag_signs })
  -- ...
end
```

------

Adding colors is a bit tricky (which is more or less by design because 'mini.statusline' and 'mini.tabline' are intentionally "colorless"), but doable. The return string of active content is essentially used as 'statusline' value, so any special statusline items (see `:h statusline`) are treated as such. Including the construct of `%#HLname#` setting the highlight group for characters to the right.

This can be utilized by setting more complicated diagnostic signs. For example, like so:

```lua
local diag_signs = {
  ERROR = '%#DiagnosticError#!%#MiniStatuslineDevinfo#',
  WARN = '%#DiagnosticWarn#?%#MiniStatuslineDevinfo#',
  INFO = '%#DiagnosticInfo#@%#MiniStatuslineDevinfo#',
  HINT = '%#DiagnosticHint#*%#MiniStatuslineDevinfo#',
}
-- ...
```

The `DiagnosticXxx` highlight groups can not play nicely with its surrounding background color, so you may have to define your own highlight groups. But that part is left for homework.

Hope this helps.
